### ISSUE TITLE
Have tables/models and dbs specified in separate modules?

### ISSUE THREAD

**Issue by rjam** [Raizlabs/DBFlow#150]
First of all, thanks for your great work with this library.

I'm trying to put all my database access logic on a separate module for reusability, meaning that I'm declaring my tables/models on that module like:

```java
@Table
public class TestModel extends BaseModel {

    @Column(columnType = Column.PRIMARY_KEY)
    String id;
}
```

But databases should be declared by the main module (imagine I have 2 apps, each would keep its own database versioning and name info). Ideally, these modules would link to the main module's database. Unfortunately this doesn't compile, with the following NPE:

```java
Caused by: java.lang.NullPointerException
        at com.raizlabs.android.dbflow.processor.definition.TableDefinition.<init>(TableDefinition.java:97)
        at com.raizlabs.android.dbflow.processor.handler.TableHandler.onProcessElement(TableHandler.java:34)
        at com.raizlabs.android.dbflow.processor.handler.BaseContainerHandler.handle(BaseContainerHandler.java:25)
        at com.raizlabs.android.dbflow.processor.model.ProcessorManager.handle(ProcessorManager.java:263)
        at com.raizlabs.android.dbflow.processor.DBFlowProcessor.process(DBFlowProcessor.java:94)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:793)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:722)
```

I'm assuming this is due to the fact that at compile time the library isn't able to access the database info from the main project? I've tried specifying the database name explicitly, but it doesn't help.. is there any way around this or am I going at this the wrong way from what you envisioned?

**Comment by agrosner** [comment]
Having them in different projects leads to a strange scenario because the annotation processor will generate or try to for each module the `GeneratedDatabaseHolder`. 

This crash is interesting as it should rather complain about it missing tables or something. 

Generally annotation processing libraries do not play nice when used across different modules. It's because it has no access into the other module's environment during processing. Not sure if other devs have figured out how to make it work. 

Your best bet is to subclass each model definition in your main module and add the annotations there. The only downside becomes if you have any kind of relationship such as 1-many.

**Comment by rjam** [comment]
> Generally annotation processing libraries do not play nice when used across different modules. It's because it has no access into the other module's environment during processing.

I see, that's what I feared :( 

> Your best bet is to subclass each model definition in your main module and add the annotations there. The only downside becomes if you have any kind of relationship such as 1-many.

This was also one of the first things I've tried but couldn't get to work. Since java doesn't allow us to override fields so all columns have to be declared on the "models module". This "models module" had the DBFlow dependency. I would then just extend the class on my main module, adding the ```@Table``` annotation. Unfortunately I got errors like the ones you mention regarding missing tables (sorry I can't be precise but the project moved on and I'm currently unable to reproduce it)

Anyway, i'm closing this since this seems to be a common limitation within annotation based libraries. Thanks for the feedback
